source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +66,111 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
+  for (uint64_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      auto host_weight = host->weight();
+      ASSERT(host_weight != 0);
+      // NOTE: Locality weight may be zero. In this case, all hosts in the locality are assigned
+      //       no load. When we build the ring below, such hosts will have no entries in the ring.
+      // TODO: When we move to C++17, change this to `locality_weights[i]` (i.e. use
+      //       std::shared_ptr::operator[]) rather than dereferencing locality_weights explicitly.
+      auto locality_weight = locality_weights == nullptr ? 1 : (*locality_weights)[i];
+      auto effective_weight = host_weight * locality_weight;
+      weighted_sum += effective_weight;
+      effective_weights[host] = effective_weight;
     }
   }
 
-  ENVOY_LOG(info, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size, hashes_per_host);
-  ring_.reserve(hosts.size() * hashes_per_host);
+  // We can't do anything sensible with no hosts.
+  if (weighted_sum == 0) {
+    return;
+  }
+
+  // Determine the valid range of replication factors given the weighted sum and ring size bounds,
+  // and the actual replication factor within that range. This may differ from the configured
+  // replication factor if provided. Note that there's some subtle but important considerations
+  // for proper consistent hashing behavior here. First, we find the absolute minimum and maximum
+  // replication factors that would fit the configured ring size bounds. These will work out to be
+  // weird fractions most of the time (e.g. 123.45 hashes per host), which doesn't make sense, of
+  // course, there's no such thing as a fractional entry on the ring...
+  double min_replication_factor =
+      static_cast<double>(min_ring_size) / static_cast<double>(weighted_sum);
+  double max_replication_factor =
+      static_cast<double>(max_ring_size) / static_cast<double>(weighted_sum);
+
+  // ... So, we find the closest whole number replication factors by rounding up the minimum and
+  // rounding down the maximum. These will be good most of the time, but there is still an edge
+  // case: suppose the user sets minimum_ring_size and maximum_ring_size both to exactly the same
+  // number, such as 1024 entries (which is something they should probably be allowed to do), and
+  // then has a weighted_sum of hosts that doesn't divide evenly into 1024, such as 5. The exact
+  // fractional replication factor computed above would be 204.8. So here we'd say that the
+  // min_whole_replication_factor would be 205 and the max would be 204, which is no good. Our
+  // policy is to try and use whole number replication factors when possible, and only use
+  // fractional replication factors when we absolutely have to.
+  const double min_whole_replication_factor = std::ceil(min_replication_factor);
+  const double max_whole_replication_factor = std::floor(max_replication_factor);
```

Comment:
What is we make them both `floor`? We'd have a slightly smaller ring potentially, but that might simplify this logic and special case. Are there degenerate examples where this would do something really bad?

Reply:
If you take `std::floor(min_replication_factor)`, you end up with a replication factor that doesn't meet the minimum ring size. A couple examples:

* You have `min_ring_size = 10` and `weighted_sum = 20`. If you divide and round down, you end up with `min_whole_replication_factor = 0`, which gives you an empty ring.
* You have `min_ring_size = 30` and `weighted_sum = 20`. If you divide and round down, you end up with `min_whole_replication_factor = 1`, which gives you a ring of size 20, smaller than the user asked for.

There might be some ways to adjust for that elsewhere in the algorithm, but my thinking was: it's important to really respect whatever ring size bounds the user set, so I'll put those constraints up front.

Reply:
We should probably then round up on the max ring.. I'm generally not a fan of this special case behavior that only kicks in under narrow conditions. It seems a violation of principle-of-least-surprise.

Reply:
Then you end up with different special-case behavior, that users might find surprising, where sometimes the ring can be larger than the limit you specified... This is easy to anticipate if all weights are 1, but I think it's harder for the user to predict what will happen with varying host and locality weights (especially if we're thinking of changing how those work here and in Maglev)...

In practical terms, I think the only place that this fractional weighting fallback will ever be noticeable is with very small replication factors. That is, a replication factor like 0.5 (where half of your hosts are taking no traffic at all, assuming weights of 1) or 1.5 (where half of your hosts are taking twice as much traffic as the other half, again assuming weights of 1). If you have a replication factor like 123.45, it's not bad at all: your effective weights are off by 0.2% in either direction, which won't be discernable at all from the error introduced by non-uniform hash distribution.

Reply:
If we do this, we'll need to make sure there is a bunch of test coverage and even then I'm still uncomfortable that we're introducing edge case handling here. @mattklein123 WDYT?

Reply:
I agree with @htuch. I would favor simplicity over doing exactly what the user says in the config, I don't think it matters that much. Can we try to simplify as much as possible? Along these lines, I haven't looked at test cases, but are you positive that all of this code will not change the ring shape assuming the user has all weights of 1 and does not specify replication factor or max ring size? We cannot force a rehash on people unless they are using the new features.

Reply:
@htuch @mattklein123 - Ok, I'm going to eliminate fractional weights, and just exceed the `maximum_ring_size` if I have to. To answer Matt's question: yes, I'm preserving the ring shape for all existing users, that's important.